### PR TITLE
Update codecs.py to CPython 3.8

### PR DESCRIPTION
**Pull request by fanninpm**
A considerable portion of the Rust code needs to be iterated upon in order for this to work correctly.

### CODE REVIEW

**Comment by fanninpm** [comment]
Nightly compilation breakage is spurious: see rust-num/num-bigint#218

File: vm/src/codecs.rs [hunk 1]
```diff
@@ -428,3 +440,255 @@ fn backslashreplace_errors(err: PyObjectRef, vm: &VirtualMachine) -> PyResult<(S
...
+    Utf32Be,
+    Utf32Le,
+    Unknown,
+}
+
+fn get_standard_encoding(encoding: &str) -> (usize, StandardEncoding) {
+    if let Some(encoding) = encoding.to_lowercase().strip_prefix("utf") {
+        let mut byte_length: usize = 0;
+        let mut standard_encoding = StandardEncoding::Unknown;
+        let encoding = match encoding.strip_prefix(|c| c == '-' || c == '_') {
```
Comment (youknowone) [inline vm/src/codecs.rs:484]:
```suggestion
        let encoding = match encoding.strip_prefix(|c| ['-', '_'].contains(c)) {
```

File: vm/src/codecs.rs [hunk 2]
```diff
@@ -428,3 +440,255 @@ fn backslashreplace_errors(err: PyObjectRef, vm: &VirtualMachine) -> PyResult<(S
...
+            if encoding.is_empty() {
+                if cfg!(target_endian = "little") {
+                    standard_encoding = StandardEncoding::Utf16Le;
+                } else if cfg!(target_endian = "big") {
+                    standard_encoding = StandardEncoding::Utf16Be;
+                }
+                match standard_encoding {
+                    StandardEncoding::Unknown => (),
+                    _ => return (byte_length, standard_encoding),
+                }
```
Comment (youknowone) [inline vm/src/codecs.rs:502]:
```suggestion
                if !matches!(standard_encoding, StandardEncoding::Unknown) {
                    return (byte_length, standard_encoding);
                }
```

Adding `#[derive(Eq)]` to `StandardEncoding` will let plain comparision operator like `if standard_encoding != StandardEncoding::Unknown`

File: vm/src/codecs.rs [hunk 3]
```diff
@@ -428,3 +440,255 @@ fn backslashreplace_errors(err: PyObjectRef, vm: &VirtualMachine) -> PyResult<(S
...
+            byte_length = 2;
+            if encoding.is_empty() {
+                if cfg!(target_endian = "little") {
+                    standard_encoding = StandardEncoding::Utf16Le;
+                } else if cfg!(target_endian = "big") {
+                    standard_encoding = StandardEncoding::Utf16Be;
+                }
+                match standard_encoding {
+                    StandardEncoding::Unknown => (),
+                    _ => return (byte_length, standard_encoding),
```
Comment (youknowone) [inline vm/src/codecs.rs:501]: another suggestion is:

```suggestion
                } else {
                    return (byte_length, standard_encoding);
                }
```

File: vm/src/codecs.rs [hunk 4]
```diff
@@ -428,3 +440,255 @@ fn backslashreplace_errors(err: PyObjectRef, vm: &VirtualMachine) -> PyResult<(S
...
+                    write!(out, "\\x{:x?}", (c >> 16)).unwrap();
+                    write!(out, "\\x{:x?}", (c >> 24)).unwrap();
+                }
+                StandardEncoding::Utf32Be => {
+                    write!(out, "\\x{:x?}", (c >> 24)).unwrap();
+                    write!(out, "\\x{:x?}", (c >> 16)).unwrap();
+                    write!(out, "\\x{:x?}", (c >> 8)).unwrap();
+                    write!(out, "\\x{:x?}", c).unwrap();
+                }
+                StandardEncoding::Unknown => unreachable!(), // NOTE: RUSTPYTHON, should've bailed out earlier
```
Comment (youknowone) [inline vm/src/codecs.rs:589]:
```suggestion
                StandardEncoding::Unknown => unreachable!("should've bailed out earlier"),
```

**Review by youknowone — APPROVED** [review]
looks great. i left a few comments about subtle things

File: vm/src/codecs.rs [hunk 5]
```diff
@@ -428,3 +440,255 @@ fn backslashreplace_errors(err: PyObjectRef, vm: &VirtualMachine) -> PyResult<(S
...
+}
+
+fn get_standard_encoding(encoding: &str) -> (usize, StandardEncoding) {
+    if let Some(encoding) = encoding.to_lowercase().strip_prefix("utf") {
+        let mut byte_length: usize = 0;
+        let mut standard_encoding = StandardEncoding::Unknown;
+        let encoding = match encoding.strip_prefix(|c| c == '-' || c == '_') {
+            Some(x) => x,
+            None => encoding,
+        };
```
Comment (DimitrisJim) [inline vm/src/codecs.rs]: You could use an `unwrap_or` here and a bit further down.

```suggestion
        let encoding = encoding.strip_prefix(|c| c == '-' || c == '_').unwrap_or(encoding);
```

File: vm/src/codecs.rs [hunk 6]
```diff
@@ -428,3 +440,255 @@ fn backslashreplace_errors(err: PyObjectRef, vm: &VirtualMachine) -> PyResult<(S
     }
     Ok((out, range.end))
 }
+
+fn namereplace_errors(err: PyObjectRef, vm: &VirtualMachine) -> PyResult<(String, usize)> {
+    if err.isinstance(&vm.ctx.exceptions.unicode_encode_error) {
```
Comment (DimitrisJim) [inline vm/src/codecs.rs:445]: nitpick: I'd handle the else clause first and return the err and, in effect, dedent this if.

File: vm/src/codecs.rs (hunk 5 above)
Comment (youknowone) [reply vm/src/codecs.rs]: does lifetime allow it? I wasn't sure about it

File: vm/src/codecs.rs [hunk 7]
```diff
@@ -428,3 +440,254 @@ fn backslashreplace_errors(err: PyObjectRef, vm: &VirtualMachine) -> PyResult<(S
...
+                write!(out, "\\x{:02x}", c_u32).unwrap();
+            }
+        }
+        Ok((out, range.end))
+    } else {
+        Err(bad_err_type(err, vm))
+    }
+}
+
+#[derive(Eq)]
```
Comment (youknowone) [inline vm/src/codecs.rs:471]: PartialEq?

File: vm/src/codecs.rs (hunk 6 above)
Comment (fanninpm) [reply vm/src/codecs.rs:445]: I was just patterning the general structure from [the CPython code](https://github.com/python/cpython/blob/dd7b816ac87e468e2fa65ce83c2a03fe1da8503e/Python/codecs.c#L959-L1051). Perhaps that improvement could be upstreamed, especially if it boosts performance.

File: vm/src/codecs.rs [hunk 8]
```diff
@@ -428,3 +440,255 @@ fn backslashreplace_errors(err: PyObjectRef, vm: &VirtualMachine) -> PyResult<(S
...
+                .strip_prefix(|c| ['-', '_'].contains(&c))
+                .unwrap_or(encoding);
+            standard_encoding = match encoding {
+                "be" => StandardEncoding::Utf32Be,
+                "le" => StandardEncoding::Utf32Le,
+                _ => StandardEncoding::Unknown,
+            }
+        }
+        return (byte_length, standard_encoding);
+    } else if encoding.to_lowercase() == "CP_UTF8" {
```
Comment (coolreader18) [inline vm/src/codecs.rs:533]: Sorry, this is merged already, but won't this always fail?

Comment (fanninpm) [reply vm/src/codecs.rs:533]: D'oh! Thanks for catching this glaring mistake. I'll patch it tonight.

```suggestion
    } else if encoding == "CP_UTF8" {
```

**Pull request merged**